File: libdata/src/main/java/com/simprints/libdata/models/firebase/fb_Person.java
package com.simprints.libdata.models.firebase;

import com.google.firebase.FirebaseApp;
import com.google.firebase.database.Exclude;
import com.google.firebase.database.Query;
import com.google.firebase.database.ServerValue;
import com.simprints.libcommon.Fingerprint;
import com.simprints.libcommon.Person;
import com.simprints.libdata.models.realm.rl_Person;
import com.simprints.libdata.tools.Utils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.simprints.libdata.tools.Routes.patientsNode;
import static com.simprints.libdata.tools.Routes.projectRef;

@SuppressWarnings({"WeakerAccess"})
public class fb_Person {

    public String patientId;
    public String userId;
    public String moduleId;
    public String androidId;
    public Long createdAt;
    private Long serverCreatedAt;
    public List<fb_Fingerprint> fingerprints;

    public fb_Person() {
    }

    public fb_Person(Person person, String userId, String androidId, String moduleId) {
        this.userId = userId;
        this.androidId = androidId;
        this.moduleId = moduleId;
        this.patientId = person.getGuid();
        this.createdAt = Utils.now().getTime();

        this.fingerprints = new ArrayList<>();
        for (Fingerprint print : person.getFingerprints()) {
            fb_Fingerprint fb_print = new fb_Fingerprint(print);
            fingerprints.add(fb_print);
        }
    }

    public fb_Person(rl_Person person) {
        this.userId = person.userId;
        this.androidId = person.androidId;
        this.patientId = person.patientId;
        this.createdAt = person.createdAt;
        this.moduleId = person.moduleId;

        this.fingerprints = new ArrayList<>();
        for (Fingerprint print : person.getLibPerson().getFingerprints()) {
            fb_Fingerprint fb_print = new fb_Fingerprint(print);
            fingerprints.add(fb_print);
        }
    }

    public java.util.Map<String, String> getServerCreatedAt() {
        return ServerValue.TIMESTAMP;
    }

    @Exclude
    public Map<String, Object> toMap() {
        HashMap<String, Object> result = new HashMap<>();
        result.put("patientId", patientId);
        result.put("userId", userId);
        result.put("moduleId", moduleId);
        result.put("androidId", androidId);
        result.put("createdAt", createdAt);
        result.put("serverCreatedAt", ServerValue.TIMESTAMP);
        result.put("fingerprints", fingerprints);
        return result;
    }

    @Exclude
    public static Query getByUser(FirebaseApp app, String apiKey, String userId) {
        return projectRef(app, apiKey).child(patientsNode()).orderByChild("userId").equalTo(userId);
    }

    @Exclude
    public static Query getByModuleId(FirebaseApp app, String apiKey, String moduleId) {
        return projectRef(app, apiKey).child(patientsNode()).orderByChild("moduleId").equalTo(moduleId);
    }
}

File: id/src/main/java/com/simprints/id/libdata/models/firebase/fb_Person.kt
package com.simprints.id.libdata.models.firebase

import com.google.firebase.FirebaseApp
import com.google.firebase.database.Exclude
import com.google.firebase.database.Query
import com.google.firebase.firestore.FieldValue
import com.google.gson.annotations.SerializedName
import com.simprints.id.libdata.models.realm.rl_Person
import com.simprints.id.libdata.tools.Routes.patientsNode
import com.simprints.id.libdata.tools.Routes.projectRef
import com.simprints.id.libdata.tools.Utils
import com.simprints.libcommon.Person
import com.simprints.libsimprints.FingerIdentifier
import java.util.*

data class fb_Person(@SerializedName("id") var patientId: String = "",
                     var userId: String = "",
                     var moduleId: String = "",
                     var androidId: String = "",
                     var createdAt: Date = Utils.now(),
                     var updatedAt: Date = Utils.now(),
                     var fingerprints: HashMap<FingerIdentifier, ArrayList<fb_Fingerprint>> = hashMapOf()) {

    constructor (person: Person,
                 userId: String,
                 androidId: String,
                 moduleId: String) : this (
        patientId = person.guid,
        userId = userId,
        androidId = androidId,
        moduleId = moduleId) {

        person.fingerprints
            .map { fb_Fingerprint(it) }
            .forEach {
                var listOfFingerprints = fingerprints[it.fingerId]
                if (listOfFingerprints == null) {
                    listOfFingerprints = arrayListOf()
                }

                listOfFingerprints.add(it)
            }
    }

    constructor (realmPerson: rl_Person) : this (
        patientId = realmPerson.patientId,
        userId = realmPerson.userId,
        moduleId = realmPerson.moduleId,
        createdAt = realmPerson.createdAt) {

        realmPerson.libPerson.fingerprints
            .map { fb_Fingerprint(it) }
            .forEach {
                var listOfFingerprints = fingerprints[it.fingerId]
                if (listOfFingerprints == null) {
                    listOfFingerprints = arrayListOf()
                }

                listOfFingerprints.add(it)
            }
    }

    @Exclude
    fun toMap(): Map<String, Any> {
        return mapOf(
            "patientId" to patientId,
            "userId" to userId,
            "moduleId" to moduleId,
            "createdAt" to createdAt,
            "syncTime" to FieldValue.serverTimestamp(),
            "fingerprints" to fingerprints)
    }

    fun getAllFingerprints(): ArrayList<fb_Fingerprint> {
        return ArrayList(fingerprints.flatMap { t -> t.value })
    }

    companion object {

        @Exclude
        fun getByUser(app: FirebaseApp, apiKey: String, userId: String): Query {
            return projectRef(app, apiKey).child(patientsNode()).orderByChild("userId").equalTo(userId)
        }

        @Exclude
        fun getByModuleId(app: FirebaseApp, apiKey: String, moduleId: String): Query {
            return projectRef(app, apiKey).child(patientsNode()).orderByChild("moduleId").equalTo(moduleId)
        }
    }
}
